fleetshard/pkg/central/reconciler/gen/main.go

```
@@ -0,0 +1,232 @@
+	}
+	defer genFile.Close()
+
+	genFile.WriteString(builder.String())
+	return nil
+}
+
+func sortResourceKeys(seen resourceMap) []schema.GroupVersionKind {
+	sorted := make([]schema.GroupVersionKind, 0, len(seen))
+	for k := range seen {
+		sorted = append(sorted, k)
+	}
+	sort.Slice(sorted, func(i, j int) bool {
+		left := sorted[i]
+		right := sorted[j]
+		return left.String() < right.String()
+	})
+	return sorted
+}
+
+func findGVKsInGeneratedFile(seen resourceMap) error {
+	if _, err := os.Stat(outFile); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+
+	file, err := os.Open(outFile)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileBytes, err := os.ReadFile(outFile)
+	if err != nil {
+		return err
+	}
+
+	lines := strings.Split(string(fileBytes), "\n")
+
+	for _, line := range lines {
+		matches := gvkRegex.FindStringSubmatch(line)
+		if len(matches) != 4 {
+			continue
+		}
+		gvk := schema.GroupVersionKind{Group: matches[1], Version: matches[2], Kind: matches[3]}
+		isAlreadyPresent := seen[gvk]
+		seen[gvk] = isAlreadyPresent
+	}
+	return nil
+}
+
+func findGVKsInChart(chartDir string, seen resourceMap) error {
+	if err := filepath.WalkDir(chartDir, func(path string, d fs.DirEntry, err error) error {
+		if d.IsDir() {
+			return nil
+		}
+
+		ext := filepath.Ext(path)
+		if ext != ".yaml" && ext != ".yml" {
+			return nil
+		}
+
+		fileBytes, err := os.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		if err := findGVKsInFile(fileBytes, seen); err != nil {
+			return err
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func splitFileIntoResources(fileBytes []byte) []string {
+	fileStr := string(fileBytes)
+	return strings.Split(fileStr, "---")
+}
+
+func findGVKsInFile(fileBytes []byte, seen resourceMap) error {
+	resources := splitFileIntoResources(fileBytes)
+	for _, resource := range resources {
+		gvk, ok, err := findResourceGVK(resource)
+		if err != nil || !ok {
```

Just to double check that this was the intention here: if the function returns ok `false` and error `nil`, it will return `nil`, so a resource without `apiVersion` or `kind` will be silently discarded.

---

Yeah that was the intention, but in light of your comment, I changed it to return an error if apiVersion/kind is not found